dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/validation/validator/enrollment/AttributeValidator.java

```diff
@@ -197,20 +178,22 @@ private void validateMandatoryAttributes(
                 enrollment,
                 E1019,
                 attrId.getIdentifierOrAttributeValue() + "=" + attrVal));
-  }
 
-  private Set<MetadataIdentifier> buildTeiAttributes(
-      TrackerBundle bundle, String trackedEntityUid) {
-    TrackerIdSchemeParams idSchemes = bundle.getPreheat().getIdSchemes();
-    return Optional.of(bundle)
-        .map(TrackerBundle::getPreheat)
-        .map(trackerPreheat -> trackerPreheat.getTrackedEntity(trackedEntityUid))
-        .map(TrackedEntity::getTrackedEntityAttributeValues)
-        .orElse(Collections.emptySet())
-        .stream()
-        .map(TrackedEntityAttributeValue::getAttribute)
-        .map(idSchemes::toMetadataIdentifier)
-        .collect(Collectors.toSet());
+    if (bundle.getStrategy(enrollment) == CREATE) {
+      // Merged attributes must contain each mandatory program attribute.
+      programAttributesMap.entrySet().stream()
+          .filter(Map.Entry::getValue) // <--- filter on mandatory flag
+          .map(Map.Entry::getKey)
+          .forEach(
+              mandatoryProgramAttribute ->
+                  reporter.addErrorIf(
+                      () -> !mergedAttributes.contains(mandatoryProgramAttribute),
```

Comment:
> Always validate that no mandatory attribute is being deleted

what if the UID is there but there is no value?

Reply:
That case is validated in `validateRequiredProperties` method.
The validator logic is quite difficult to follow but I didn't want to refactor the validator before I managed to backport the fix.
I will do it in a following PR